jerry-main/main-unix.c

```diff
@@ -629,99 +629,148 @@ main (int argc,
+            {
+              ret_value = jerry_create_error (JERRY_ERROR_COMMON,
+                                              (jerry_char_t *) "Connection aborted before source arrived.");
+            }
+
+            if (receive_status == JERRY_DEBUGGER_SOURCE_END)
+            {
+              jerry_port_log (JERRY_LOG_LEVEL_DEBUG, "No more client source.\n");
+            }
+
+            jerry_release_value (run_result);
           }
+          while (receive_status == JERRY_DEBUGGER_SOURCE_RECEIVED);
 
-          if (receive_status == JERRY_DEBUGGER_SOURCE_END)
+          if (receive_status != JERRY_DEBUGGER_CONTEXT_RESET_RECEIVED)
           {
-            jerry_port_log (JERRY_LOG_LEVEL_DEBUG, "No more client source.\n");
+            break;
           }
 
-          jerry_release_value (run_result);
-        }
-        while (receive_status == JERRY_DEBUGGER_SOURCE_RECEIVED);
+          jerry_cleanup ();
 
-        if (receive_status != JERRY_DEBUGGER_CONTEXT_RESET_RECEIVED)
-        {
-          break;
+          jerry_init (flags);
+          jerry_debugger_init (debug_port);
+
+          register_js_function ("assert", jerryx_handler_assert);
+          register_js_function ("gc", jerryx_handler_gc);
+          register_js_function ("print", jerryx_handler_print);
+
+          ret_value = jerry_create_undefined ();
         }
+      }
 
-        jerry_cleanup ();
+    }
 
-        jerry_init (flags);
-        jerry_debugger_init (debug_port);
+    bool restart = false;
 
-        register_js_function ("assert", jerryx_handler_assert);
-        register_js_function ("gc", jerryx_handler_gc);
-        register_js_function ("print", jerryx_handler_print);
+    if (jerry_is_feature_enabled (JERRY_FEATURE_DEBUGGER))
+    {
+      if (jerry_value_is_abort (ret_value))
+      {
+        jerry_value_t abort_value = jerry_get_value_from_error (ret_value, false);
+        if (jerry_value_is_string (abort_value))
+        {
+          jerry_char_t str_buf[6];
+          jerry_value_t str_val = jerry_value_to_string (abort_value);
+          jerry_size_t str_size = jerry_get_string_size (str_val);
+          jerry_size_t str_end = jerry_string_to_char_buffer (str_val, str_buf, str_size);
+          assert (str_end == str_size);
+          str_buf[str_end] = 0;
+
+          if (strcmp ("r353t", (const char *) (str_buf)) == 0)
+          {
+            jerry_release_value (ret_value);
+            restart = true;
+          }
 
-        ret_value = jerry_create_undefined ();
+          jerry_release_value (str_val);
+        }
+
+        jerry_release_value (abort_value);
       }
     }
 
+    if (!restart)
+    {
+      break;
+    }
+
+    jerry_cleanup ();
+
+    jerry_init (flags);
+    jerry_debugger_init (debug_port);
+
+    register_js_function ("assert", jerryx_handler_assert);
+    register_js_function ("gc", jerryx_handler_gc);
+    register_js_function ("print", jerryx_handler_print);
```

Comment:
This looks like a code duplication (init engine). Would be good to move it into a static function.